docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,405 @@
+      configMap:
+        namespace: foo
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+```
+
+If a function is deleted upstream and not changed on the local, it will be deleted on local.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+
+Expected output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+```
+
+Same function declared multiple times: If the same function is declared multiple 
+times with the same input type(configMap/configPath), order is used as a tie-breaker 
+to identify the function, which means the functions are merged based on their order
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: foo
+        put-value: bar
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: abc
+        put-comment: ${some-setter-name}
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: foo
+        put-value: bar-new
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: abc
+        put-comment: ${updated-setter-name}
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: foo
+        put-value: bar
+    - image: gcr.io/kpt-fn/set-labels:v0.1
+      configMap:
+        app: db
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: abc
+        put-comment: ${some-setter-name}
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: YOUR_TEAM
+        put-value: my-team
+
+Expected output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: foo
+        put-value: bar-new
+    - image: gcr.io/kpt-fn/set-labels:v0.1
+      configMap:
+        app: db
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: abc
+        put-comment: ${updated-setter-name}
+    - image: gcr.io/kpt-fn/search-replace:v0.1
+      configMap:
+        by-value: YOUR_TEAM
+        put-value: my-team
+```
+
+Merging selectors is difficult as there is no identity. If both upstream and 
```

Comment:
This is in theory an easier problem, since the behavior is not dependent on the ordering. Also, more than one entry with the same values doesn't make sense, so I would think that treating the combination of the kind and name as the merge key would work?

Reply:
kind and name are not mandatory selector options. All selector options are optional and completely independent. It is difficult to have the merge key here.